src/gordon_gcp/plugins/enricher.py

```diff
@@ -49,11 +57,96 @@ class GCEEnricher:
+        self._logger = logging.getLogger('')
+
+    async def _poll_for_ext_ip(self, resource_name, msg_logger):
+        exception = None
+        backoff = 2
+        base_url = f'https://compute.googleapis.com/compute/v1/{resource_name}'
+
+        for attempt in range(1, self.config['retries'] + 1):
+            try:
+                msg_logger.debug(f'Attempt {attempt}: fetching {base_url}')
+                instance_data = await self._http_client.get_json(base_url)
+                return instance_data['networkInterfaces'][0][
+                    'accessConfigs'][0]['natIP']
+            except exceptions.GCPHTTPError as e:
+                exception = e
+                break
+            except (KeyError, IndexError) as e:
+                exception = e
+                if attempt == self.config['retries']:
+                    continue
+                await asyncio.sleep(backoff)
+                backoff = backoff ** 2
+
+        msg = (f'Could not get external IP for {resource_name}: '
+               f'{exception.__class__.__name__}: {exception}')
+        raise exceptions.GCPGordonError(msg)
+
+    async def _create_A_rrecord(self, data, msg_logger):
+        external_ip = await self._poll_for_ext_ip(data['resourceName'],
+                                                  msg_logger)
+        dns_zone = self.config['dns_zone']
+        hostname = data['resourceName'].split('/')[-1]
+
+        name = '.'.join([hostname, dns_zone])
+        msg_logger.debug(f'Creating A record: {name} -> {external_ip}')
+        return {
+            'name': name,
+            'rrdatas': [external_ip],
+            'type': 'A',
+            'ttl': self.config['default_ttl']
+        }
+
+    async def process(self, event_message):
+        """
+        Enrich message with extra context and send it to the publisher.
+
+        When a message is successfully processed, it is passed to the
+        :obj:`self.success_channel`. However, if there is a problem
+        during processing, the message is passed to the
+        :obj:`self.error_channel`.
+
+        Args:
+            event_message (.GEventMessage): message requiring
+                additional information.
+        """
+        msg_logger = _utils.GEventMessageLogger(
+            self._logger, {'msg_id': event_message.msg_id})
+        data = event_message.data
+        try:
+            a_record = await self._create_A_rrecord(data, msg_logger)
+        except exceptions.GCPGordonError as e:
+            msg = ('Encountered error while enriching message, sending message'
+                   f' to error channel: {e}.')
+            msg_logger.warning(msg)
+            event_message.append_to_history(msg, self.phase)
+            await self.error_channel.put(event_message)
+            return
+        msg_logger.debug(f'Created resource record {a_record}')
+        event_message.data['resourceRecords'].append(a_record)
+        added_a_records = event_message.data['resourceRecords']
```

Comment:
So, hm, how would someone (us, too) extend this to process more record types? It doesn't seem like one would be able to without rewriting all of `process`. Maybe, `process` could iterate over some registration of record types that are mapped to their respective methods. Something like (I have no idea if this would work, and this is just the first approach I thought of, there may be better ones):

```python
# original
class GCEEnricher:
    RECORD_HANDLERS = {
        'A': '_create_A_rrecord',  # string name of method
    }
    async def process(self, msg):
        # prep stuff
        for rtype, method in RECORD_HANDLERS.items():
            records = await getattr(self, method)(data, msg_logger)
        # other stuff

# extended
class MyEnricher(GCEEnricher):
    RECORD_HANDLERS.update({ 
        'CNAME': '_create_cname_rrecord',
    }
```

Reply:
Thanks for pointing this out. I went with the approach we're using in the Janitor code eg. overwritable methods. Users that want more records can override the `_create_rrecords` method to call other, new methods for creating specific records. I'm not 100% sure this is the most comfortable way, but it seems the simplest.